x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportUnfollowAction.java

```diff
@@ -49,26 +78,131 @@ protected AcknowledgedResponse newResponse() {
+                            }
+
+                        },
+                        numberOfShards,
+                        Collections.emptyList());
+                for (int i = 0; i < numberOfShards; i++) {
+                    final ShardId followerShardId = new ShardId(indexMetaData.getIndex(), i);
+                    final ShardId leaderShardId = new ShardId(leaderIndex, i);
+                    final AtomicInteger tryCounter = new AtomicInteger(1);
+                    removeRetentionLeaseForShard(
+                            leaderShardId,
+                            retentionLeaseId,
+                            remoteClient,
+                            ActionListener.wrap(
+                                    groupListener::onResponse,
+                                    e -> handleException(
+                                            followerShardId,
+                                            retentionLeaseId,
+                                            leaderShardId,
+                                            remoteClient,
+                                            tryCounter,
+                                            groupListener,
+                                            e)));
+                }
             }
+
+            private void handleException(
+                    final ShardId followerShardId,
+                    final String retentionLeaseId,
+                    final ShardId leaderShardId,
+                    final Client remoteClient,
+                    final AtomicInteger tryCounter,
+                    final GroupedActionListener<RetentionLeaseActions.Response> groupListener,
+                    final Exception e) {
+                final Throwable cause = ExceptionsHelper.unwrapCause(e);
+                if (cause instanceof RetentionLeaseNotFoundException) {
+                    // treat as success
+                    groupListener.onResponse(new RetentionLeaseActions.Response());
+                } else if (TransportActions.isShardNotAvailableException(e) && tryCounter.get() < 16) {
+                    logger.trace(new ParameterizedMessage(
+                                    "{} leader primary shard {} not available on try [{}] while removing retention lease [{}]",
+                                    followerShardId,
+                                    leaderShardId,
+                                    tryCounter.get(),
+                                    retentionLeaseId),
+                            e);
+                    tryCounter.incrementAndGet();
+                    removeRetentionLeaseForShard(
+                            leaderShardId,
+                            retentionLeaseId,
+                            remoteClient,
+                            ActionListener.wrap(
+                                    groupListener::onResponse,
+                                    // TODO: should this should exponentially backoff, or should we simply never retry?
```

Comment:
@martijnvg @dnhatn I opened this PR as WIP for discussion on this point. I am curious to hear your thoughts. Note that failure is a definite possibility since a common use case for the unfollow API will be when the leader is unavailable.

Reply:
I'm leaning towards not retrying. However I think that we should add a field in the response indication whether retention lease removal was successful. (`retention_lease_removal` boolean field?)

Reply:
(my motivation for not retrying is, that because it is a common scenario and therefore retrying would be futile to begin with).

Reply:
Yes, that's my inclination as well.

@dnhatn Is this okay with you too?

Reply:
I am + 1 to "not retry here" because a retention lease will be automatically expired. We need to make sure ILM works well in this scenario.